cpp/src/Glacier2Lib/NullPermissionsVerifier.cpp

```diff
@@ -26,114 +26,88 @@ class NullSSLPermissionsVerifier : public Glacier2::SSLPermissionsVerifier
 {
 public:
 
-    virtual bool
-    authorize(Glacier2::SSLInfo, string&, const Ice::Current&) const
+    bool authorize(Glacier2::SSLInfo, string&, const Ice::Current&) const final
     {
         return true;
     }
 };
 
-class Init
+void setupObjectAdapter(
+    const Ice::CommunicatorPtr& communicator,
+    Ice::ObjectAdapterPtr& adapter)
 {
-public:
-
-    Init(const CommunicatorPtr&, const string&, const vector<string>&);
-
-private:
-
-    string checkPermissionVerifier(const string&);
-    void createObjects();
+    if (adapter == nullptr)
+    {
+        // Create collocated object adapter
+        adapter = communicator->createObjectAdapter("");
+        adapter->activate();
+    }
+}
 
-    const CommunicatorPtr _communicator;
-    ObjectAdapterPtr _adapter;
+}
 
-    Identity _nullPVId;
-    Identity _nullSSLPVId;
-};
+namespace Glacier2Internal
+{
 
-Init::Init(const CommunicatorPtr& communicator, const string& category, const vector<string>& props) :
-    _communicator(communicator)
+void
+setupNullPermissionsVerifier(
+    const CommunicatorPtr& communicator,
+    const string& category,
+    const vector<string>& permissionsVerifierPropertyNames)
 {
-    _nullPVId.name = "NullPermissionsVerifier";
-    _nullPVId.category = category;
+    const Ice::Identity nullPermissionsVerifierId { "NullPermissionsVerifier", category };
+    const Ice::Identity nullSSLPermissionsVerifierId { "NullSSLPermissionsVerifier",  category };
 
-    _nullSSLPVId.name = "NullSSLPermissionsVerifier";
-    _nullSSLPVId.category = category;
+    Ice::PropertiesPtr properties = communicator->getProperties();
+    ObjectAdapterPtr adapter;
 
-    Ice::PropertiesPtr properties = _communicator->getProperties();
-    for(vector<string>::const_iterator p = props.begin(); p != props.end(); ++p)
+    shared_ptr<Glacier2::PermissionsVerifier> nullPermissionsVerifier;
+    shared_ptr<Glacier2::SSLPermissionsVerifier> nullSSLPermissionsVerifier;
+
+    for (const auto& propertyName : permissionsVerifierPropertyNames)
```

Comment:
I don't understand this code.

Reply:
Reworked the code a bit:

We called this method like:

```
vector<string> verifierProperties;
    verifierProperties.push_back("Glacier2.PermissionsVerifier");
    verifierProperties.push_back("Glacier2.SSLPermissionsVerifier");

    try
    {
        Glacier2Internal::setupNullPermissionsVerifier(communicator(), instanceName, verifierProperties);
    }
    catch(const std::exception& ex)
    {
        ServiceError err(this);
        err << "unable to setup null permissions verifier:\n" << ex;
        return false;
    }
```

Here we check if any of the properties require the null permissions verifier, and create it if required.

I remove the `ProxyParseException` handling code. Updated the callers to report it as an error.

Reply:
Now that I see the code, I would not pass these properties. setupNullPermissionsVerifier should just read these properties - we don't need a `for` loop, it's just confusing.

Reply:
```cpp
string permissionVerifierValue = properties->getProperty("Glacier2.PermissionVerifier");
if proxy's identity == nullPermissionsVerifierId ...

string sslPermissionVerifierValue = properties->getProperty("Glacier2.SslPermissionVerifier");
if proxy's identity == nullSslPermissionsVerifierId ...
```

no loop

Reply:
There is 4 properties in IceGrid case:

```
vector<string> verifierProperties;
    verifierProperties.push_back("IceGrid.Registry.PermissionsVerifier");
    verifierProperties.push_back("IceGrid.Registry.SSLPermissionsVerifier");
    verifierProperties.push_back("IceGrid.Registry.AdminPermissionsVerifier");
    verifierProperties.push_back("IceGrid.Registry.AdminSSLPermissionsVerifier");

    Glacier2Internal::setupNullPermissionsVerifier(_communicator, _instanceName, verifierProperties);
```